rest/security-rest-cxfwrapper/src/main/java/org/codice/ddf/cxf/SecureCxfClientFactoryBuilder.java

```diff
@@ -0,0 +1,351 @@
+        httpConduit.setTlsClientParameters(tlsParams);
+    }
+
+    public class SecureCxfClientFactory<T> {
+
+        private Client cxfClient;
+
+        private SecureCxfClientFactory(Client cxfClient) {
+            this.cxfClient = cxfClient;
+        }
+
+        /**
+         * Clients produced by this method will be secured with basic authentication
+         * (if a username and password were provided), two-way ssl,
+         * and the provided security subject.
+         * <p/>
+         * The returned client should NOT be reused between requests!
+         * This method should be called for each new request in order to ensure
+         * that the security subject is up-to-date each time.
+         *
+         * @see SecureCxfClientFactoryBuilder
+         */
+        public T getClientForSubject(Subject subject) throws SecurityServiceException {
+            String asciiString = cxfClient.getBaseURI().toASCIIString();
+            if (!StringUtils.startsWithIgnoreCase(asciiString, "https")) {
+                throw new SecurityServiceException(
+                        "Cannot secure non-https connection " + asciiString);
+            }
+
+            WebClient newClient = WebClient.fromClient(cxfClient);
+
+            if (subject instanceof ddf.security.Subject) {
+                RestSecurity.setSubjectOnClient((ddf.security.Subject) subject, newClient);
+            } else {
+                throw new SecurityServiceException("Not a ddf subject " + subject);
+            }
+            WebClient.getConfig(newClient).getRequestContext()
+                    .put(org.apache.cxf.message.Message.MAINTAIN_SESSION, Boolean.TRUE);
+
+            return (T) newClient;
+        }
+
+        /**
+         * Clients produced by this method will be secured with basic authentication
+         * (if a username and password were provided), two-way ssl,
+         * and the system security token (x509 cert).
+         */
+        public T getClientForSystem() throws SecurityServiceException {
+            String asciiString = cxfClient.getBaseURI().toASCIIString();
+            if (!StringUtils.startsWithIgnoreCase(asciiString, "https")) {
+                throw new SecurityServiceException(
+                        "Cannot secure non-https connection " + asciiString);
+            }
+
+            if (StringUtils.isBlank(getStsClientConfiguration().getAddress())) {
+                throw new SecurityServiceException(
+                        "STSClientConfiguration is either null or its address is blank - assuming no STS Client is configured, so no SAML assertion will get generated.");
+            }
+            WebClient newClient = WebClient.fromClient(cxfClient);
+            ClientConfiguration clientConfig = WebClient.getConfig(newClient);
+            Bus clientBus = clientConfig.getBus();
+            STSClient stsClient = configureSTSClient(clientBus);
+
+            try {
+                SecurityToken securityToken = stsClient
```

Comment:
Hmm, I'm wondering if a better way to do this would be to call the SecurityManager.getSubject() with a PKI authtoken. That would allow you to use the STSRealm which duplicates a lot of the STS code in here. Not sure if that would work exactly like you want it, but it would make this a bit cleaner as it would follow the security framework that is already established. I think I would like to see most of the STS communication going through that route rather than ad-hoc requests. That would definitely help if a bug is found with how we're doing STS calls as you would only need to fix it in that one area. Once you get back the subject you could then use RestSecurity to set it on the client. Once again, not sure if it would work, but it's worth trying to see.